### PR TITLE
CON-207 Display backend logs through http

### PR DESCRIPTION
**Pull request by IssaIan**
#### Description
This PR enables the  Super Admin to view backend logs through a HTTP get request. The most recent logs appear at the top. Currently, we have to wait for the dev ops team to provide us with the log file on the staging app whenever debugging is required. This features enables one to view the logs immediately thus reducing the debugging time.

#### How this should be manually tested
Clone the repo: git clone https://github.com/andela/mrm_api.git.
Setup the project as per the README.md.
Checkout to branch story/CON-207-display-backend-logs.
On your Insomnia, go to url `localhost:8000/logs` and send a `GET REQUEST` using a `Super Admin token`. 
 
#### Type of change
- [ ] Bug fix (nonbreaking change which fixes an issue)
- [x] New feature (nonbreaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

#### How Has This Been Tested?
Please describe the tests that you ran to verify your changes
- [x] End to end
- [x] Integration

#### Checklist:
- [x] My code follows the style guide for this project
- [x] I have linted my code prior to submission
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

#### Screenshots
image

#### Jira
[CON-207](https://andela-apprenticeship.atlassian.net/jira/software/projects/CON/boards/15/backlog?label=backend&selectedIssue=CON-207)

### CODE REVIEW

File: utilities/file_reader.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+def generate(file):
```
Comment (rajeman) [inline utilities/file_reader.py]: @IssaIan you can change this function name to something that gives more information about what it does.

File: app.py [hunk 2]
```diff
@@ -27,6 +27,12 @@ def create_app(config_name):
     def index():
         return render_template('index.html')
 
+    @app.route("/logs", methods=['GET'])
+    @Auth.user_roles('Super_Admin', 'REST')
+    def logs():
+        file = 'mrm.err.log'
```
Comment (rajeman) [inline app.py]: variable `file` should be more descriptive.

File: utilities/file_reader.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
...
+    for line in reversed(list(open(file))):
+        yield line
```
Comment (rajeman) [inline utilities/file_reader.py]: It is good that the logs are being displayed `most recent first`. Try to group them by the `timestamp` that precedes every log.

Comment (IssaIan) [reply utilities/file_reader.py]: The most recent logs are displayed first. This is what was requested of me @rajeman.

Comment (IssaIan) [reply utilities/file_reader.py]: Furthermore, not every line in the log file is preceded by a timestamp.

Comment (rajeman) [reply utilities/file_reader.py]: You can look through the file again and verify every log is preceded by a timestamp.  If you need clarification, kindly reach out to me.

Comment (IssaIan) [reply utilities/file_reader.py]: @rajeman from the following screenshot, you can see that the some lines do not contain the preceding timestamp.
image

**Comment by IssaIan** [comment]
> @IssaIan well done. any reason why this feature is just for `super_admin's`

@vic3king that was what was requested by @joshuaocero. He suggested that this functionality be restricted to the super admin only.

**Comment by joshuaocero** [comment]
> > @IssaIan well done. any reason why this feature is just for `super_admin's`
> 
> @vic3king that was what was requested by @joshuaocero. He suggested that this functionality be restricted to the super admin only.

Yes I did. We would not want our stakeholders being able to see our log files.

File: utilities/file_reader.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import re
+
+
+def read_file(log_file):
```
[the remark this answers is not in the document]
Comment (IssaIan) [reply utilities/file_reader.py]: Done @mwinel

File: tests/base.py [hunk 5]
```diff
@@ -384,3 +390,16 @@ def func_wrapper(self):
         user.roles.append(user_role)
         db_session().commit()
     return func_wrapper
+
+
+def change_admin_user_role_to_super_admin(func):
+    def func_wrapper(self):
+        user_role = Role(role='Super_Admin')
```
Comment (joshuaocero) [inline tests/base.py]: This role is now `Super Admin` not `Super_Admin`

File: app.py [hunk 6]
```diff
@@ -27,6 +27,12 @@ def create_app(config_name):
     def index():
         return render_template('index.html')
 
+    @app.route("/logs", methods=['GET'])
+    @Auth.user_roles('Super_Admin', 'REST')
```
Comment (joshuaocero) [inline app.py]: This role is now `Super Admin` not `Super_Admin`

**Comment by vic3king** [comment]
> > > @IssaIan well done. any reason why this feature is just for `super_admin's`
> > 
> > 
> > @vic3king that was what was requested by @joshuaocero. He suggested that this functionality be restricted to the super admin only.
> 
> Yes I did. We would not want our stakeholders being able to see our log files.

Coll. That makes sense.

File: tests/base.py (hunk 5 above)
Comment (IssaIan) [reply tests/base.py]: Done @joshuaocero

File: utilities/file_reader.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
```
Comment (rajeman) [inline utilities/file_reader.py]: @IssaIan you can change the function name to `read_log_file`.

File: utilities/file_reader.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    """
+    pattern, error_details = re.compile(r'(\d+-\d+-\d+ \d+:\d+:\d+)'), []
+    for line in reversed(list(open(log_file, 'r'))):
+        if pattern.search(line):  # check whether line contains timestamp
+            yield line
+            for detail in reversed(error_details):
+                yield detail
+            error_details.clear()
+        else:
+            error_details.append(line)  # add error log detils to a list
```
Comment (rajeman) [inline utilities/file_reader.py]: correct the spelling of `details`

File: utilities/file_reader.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import re
+
+
+def read_file(log_file):
+    """
+    Function that accepts log file and opens it.
+    Opens the file in reverse order, iterates through
+    every line to return the logs.
+    """
+    pattern, error_details = re.compile(r'(\d+-\d+-\d+ \d+:\d+:\d+)'), []
```
Comment (rajeman) [inline utilities/file_reader.py]: change `pattern` to `timestamp_pattern` and make the variable declarations on different lines.

File: utilities/file_reader.py (hunk 8 above)
Comment (IssaIan) [reply utilities/file_reader.py]: Done.

File: utilities/file_reader.py (hunk 7 above)
Comment (IssaIan) [reply utilities/file_reader.py]: Done @rajeman

File: utilities/file_reader.py (hunk 9 above)
Comment (IssaIan) [reply utilities/file_reader.py]: Done.

**Pull request merged**